sealtk/gui/PlayerTool.cpp

```diff
@@ -0,0 +1,105 @@
+/* This file is part of SEAL-TK, and is distributed under the OSI-approved BSD
+ * 3-Clause License. See top-level LICENSE file or
+ * https://github.com/Kitware/seal-tk/blob/master/LICENSE for details. */
+
+#include <sealtk/gui/PlayerTool.hpp>
+#include <sealtk/gui/Player.hpp>
+
+#include <QPair>
+#include <QStack>
+#include <QString>
+#include <QVariant>
+
+namespace sealtk
+{
+
+namespace gui
+{
+
+// ============================================================================
+class PlayerToolPrivate
+{
+public:
+  PlayerToolPrivate(Player* player);
+
+  Player* player;
+  QStack<QPair<QString, QVariant>> pushedProperties;
+};
+
+// ----------------------------------------------------------------------------
+PlayerToolPrivate::PlayerToolPrivate(Player* player)
+  : player{player}
+{
+}
+
+// ----------------------------------------------------------------------------
+QTE_IMPLEMENT_D_FUNC(PlayerTool)
+
+// ----------------------------------------------------------------------------
+PlayerTool::PlayerTool(Player* parent)
+  : d_ptr{new PlayerToolPrivate{parent}},
+    QObject{parent}
+{
+}
+
+// ----------------------------------------------------------------------------
+PlayerTool::~PlayerTool()
+{
+}
+
+// ----------------------------------------------------------------------------
+Player* PlayerTool::player() const
+{
+  QTE_D();
+  return d->player;
+}
+
+// ----------------------------------------------------------------------------
+void PlayerTool::activate()
+{
+}
+
+// ----------------------------------------------------------------------------
+void PlayerTool::deactivate()
+{
+  QTE_D();
+
+  while (!d->pushedProperties.empty())
+  {
+    auto const pair = d->pushedProperties.pop();
+    auto const name = pair.first.toStdString();
```

Comment:
See the definition of `pushProperty()` below. I needed a string type capable of holding property names to use as the key for the `QMap`, and used `QString`. Would `std::string` be better for this purpose?

Reply:
> See the definition of `pushProperty()` below

Hmm, now that you mention it, that isn't using the right conversion either.

> Would `std::string` be better for this purpose?

Possibly... or QByteArray (roughly the same difference).

If you have *actual* (i.e. textual) strings, the "correct" conversions are usually:

| You Have | You Want | Use |
| -------- | -------- | --- |
| std::string | QString | `qtString` (from [`qtStlUtil.h`](../../qtextensions/tree/master/core/qtStlUtil.h)) |
| QString | std::string | `stdString` (from [`qtStlUtil.h`](../../qtextensions/tree/master/core/qtStlUtil.h)) |
| `char const*` | QString | [`QString::fromLocal8Bit`](https://doc.qt.io/qt-5/qstring.html#fromLocal8Bit) |
| QString | `char const*` | [`qPrintable`](https://doc.qt.io/qt-5/qtglobal.html#qPrintable) |

This will DTRT in case the string contains non-Latin1 characters or the system locale is not UTF-8. (The former was a problem in Qt4 that may be partially fixed in Qt 5. The latter is likely to matter on Windows; thankfully it is becoming increasingly unlikely elsewhere.)

Aside from `toStdString` assuming UTF-8 (at least in Qt5; IIRC in Qt4 it was latin1!), you're paying for an extra and unnecessary conversion to `std::string` compared to `qPrintable`. Of course, you could replace `toStdString` with `toLocal8Bit` and get the same effect.

However, since on further reflection your inputs are *always* `char const*`, there is no need and little benefit to paying for textual conversion; a bag-of-bytes container — `QByteArray` or `std::string` — would be fine.